### PR TITLE
Fix various issues

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

This resolves ~19 of our current issues. Working on a few more but I'll put in separate PR for that.

### Types of changes:

- [x] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### CODE REVIEW

File: tests/Unit/ErrorHandling/Test_FloatingPointExceptions.cpp [hunk 1]
```diff
@@ -13,8 +13,8 @@ SPECTRE_TEST_CASE("Unit.ErrorHandling.FloatingPointExceptions.Invalid",
                   "[ErrorHandling][Unit]") {
   ERROR_TEST();
   enable_floating_point_exceptions();
-  double x = -1.0;
-  double invalid = sqrt(x);
+  volatile double x = -1.0;
```
Comment (osheamonn) [inline tests/Unit/ErrorHandling/Test_FloatingPointExceptions.cpp:16]: Is this change related to the mac travis build? If not it should be different commit or at least described in this commit message.

Comment (nilsdeppe) [reply tests/Unit/ErrorHandling/Test_FloatingPointExceptions.cpp:16]: It is because they were failing until this change

Comment (nilsdeppe) [reply tests/Unit/ErrorHandling/Test_FloatingPointExceptions.cpp:16]: In release mode

File: tests/Unit/Utilities/Test_TypeTraits.cpp [hunk 2]
```diff
@@ -551,3 +551,40 @@ static_assert(cpp17::is_same_v<
...
+              "Failed testing remove_cvref_wrap");
+static_assert(
+    cpp17::is_same_v<tt::remove_cvref_wrap_t<const volatile int&&>, int>,
+    "Failed testing remove_cvref_wrap");
+static_assert(
+    cpp17::is_same_v<tt::remove_cvref_wrap_t<std::reference_wrapper<const int>>,
+                     int>,
+    "Failed testing remove_cvref_wrap");
+static_assert(
+    cpp17::is_same_v<tt::remove_cvref_wrap_t<std::reference_wrapper<int>>, int>,
```
Comment (nilsdeppe) [inline tests/Unit/Utilities/Test_TypeTraits.cpp:588]: add test of `const std::reference_wrapper<int>`

Comment (nilsdeppe) [reply tests/Unit/Utilities/Test_TypeTraits.cpp:588]: add test for `std::reference_wrapper<int*>`

File: docs/MainSite/Installation.md [hunk 3]
```diff
@@ -46,7 +46,11 @@ See LICENSE.txt for details.
 
 A [Docker](https://www.docker.com/) image is available from
 [DockerHub](https://hub.docker.com/r/sxscollaboration/spectrebuildenv/) and can
-be used to build SpECTRE on a personal machine.
+be used to build SpECTRE on a personal machine. If you have SELinux active
+on your system you must figure out how to enable sharing files with the host
+OS. If you receive errors that you do not have permission to access a shared
+directory it is likely that your system has SELinux enabled. One option is to
+disable SELinux at the expensive of reducing the security of your system.
```
Comment (nilsdeppe) [inline docs/MainSite/Installation.md]: expensive -> expense

File: .travis.yml [hunk 4]
```diff
@@ -95,6 +95,14 @@ matrix:
...
     dist: trusty
     env: BUILD_TYPE=Release BUILD_CODE=true
+  - os: linux
+    compiler: gcc-6
+    dist: trusty
+    env: BUILD_TYPE=Debug BUILD_CODE=true COMPILER=gcc-6
+  - os: linux
+    compiler: gcc-6
+    dist: trusty
+    env: BUILD_TYPE=Release BUILD_CODE=true
```
Comment (nilsdeppe) [inline .travis.yml]: COMPILER=gcc-6

File: tests/Unit/PointwiseFunctions/MathFunctions/Test_TensorProduct.cpp [hunk 5]
```diff
@@ -16,8 +16,8 @@
 #include "Domain/CoordinateMaps/CoordinateMap.hpp"
 #include "Domain/CoordinateMaps/ProductMaps.hpp"
 #include "Domain/LogicalCoordinates.hpp"
-#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp"
 #include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.tpp"
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/MathFunctions/Test_TensorProduct.cpp]: move .tpp to bottom

File: src/DataStructures/Variables.hpp [hunk 6]
```diff
@@ -44,7 +44,19 @@ class Variables;
...
+ * value must be a `Tensor<DataVector, ...>`, a `static constexpr
+ * db::DataBoxString` variable named `label`, and must derive off of
+ * `db::DataBoxTag`. In general, they should be DataBoxTags that are not compute
+ * items. For example,
+ *
+ * \snippet Test_Variables.cpp simple_variables_tag
+ *
+ * Prefix tags can also be stored and their format is:
+ *
+ * * \snippet Test_Variables.cpp prefix_variables_tag
```
Comment (nilsdeppe) [inline src/DataStructures/Variables.hpp]: RIP

File: cmake/SetupCppCheck.cmake [hunk 7]
```diff
@@ -6,19 +6,37 @@ if(NOT CPPCHECK_FOUND AND NOT CPPCHECK_EXECUTABLE)
 endif()
 
 if(CPPCHECK_EXECUTABLE)
+  file(
```
Comment (nilsdeppe) [inline cmake/SetupCppCheck.cmake]: explain why can't just use the return value from cppcheck

File: cmake/SetupCppCheck.cmake [hunk 8]
```diff
@@ -6,19 +6,37 @@ if(NOT CPPCHECK_FOUND AND NOT CPPCHECK_EXECUTABLE)
...
+    WRITE
+    ${CMAKE_BINARY_DIR}/.cppcheck_run.sh
+    "#!/bin/bash -e\n"
+    "cppcheck "
+    "--inline-suppr "
```
Comment (nilsdeppe) [inline cmake/SetupCppCheck.cmake]: Indent these by 2 to make more legible

File: cmake/SetupCppCheck.cmake [hunk 9]
```diff
@@ -6,19 +6,37 @@ if(NOT CPPCHECK_FOUND AND NOT CPPCHECK_EXECUTABLE)
...
       COMMAND
-      ${CPPCHECK_EXECUTABLE}
-      --inline-suppr
-      --template=\"[{severity}][{id}] {message} {callstack}\"
-      --enable=warning,performance,portability,information,missingInclude
-      --project=${CMAKE_BINARY_DIR}/compile_commands.json
-      --suppressions-list=${CMAKE_SOURCE_DIR}/tools/SuppressionsCppCheck.txt
-      -j2
-
+      bash -c ${CMAKE_BINARY_DIR}/.cppcheck_run.sh
```
Comment (nilsdeppe) [inline cmake/SetupCppCheck.cmake]: remove bash -c

File: tools/ClangTidyHash.sh [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+# Go to build directory and then run clang-tidy, writing output to file
+cd ${BUILD_DIR}
+NO_ERROR=0
+
+for FILENAME in $MODIFIED_FILES
+do
+    printf "\n\nRunning clang-tidy on file $FILENAME\n"
+    make clang-tidy FILE=$SOURCE_DIR/${FILENAME}
+    NO_ERROR=$? || $NO_ERROR
```
Comment (nilsdeppe) [inline tools/ClangTidyHash.sh]: if $? is false, set NO_ERROR to false

File: .travis/RunClangTidy.sh [hunk 11]
```diff
@@ -56,6 +69,7 @@ echo ''
 
 for FILENAME in $MODIFIED_FILES
 do
+    printf "\n\nRunning clang-tidy on file $FILENAME\n"
     make clang-tidy FILE=/work/spectre/${FILENAME} 2>&1 >> ${CLANG_TIDY_OUTPUT}
```
Comment (nilsdeppe) [inline .travis/RunClangTidy.sh]: move 2>&1 to end, and test

File: tests/Unit/DataStructures/Test_Variables.cpp [hunk 12]
```diff
@@ -12,9 +12,11 @@
 #include "tests/Unit/TestHelpers.hpp"
 
 namespace VariablesTestTags_detail {
+/// [simple_variables_tag]
 struct vector : db::DataBoxTag {
   using type = tnsr::I<DataVector, 3, Frame::Grid>;
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_Variables.cpp]: add a label

File: cmake/SetupCppCheck.cmake [hunk 13]
```diff
@@ -6,17 +6,21 @@ if(NOT CPPCHECK_FOUND AND NOT CPPCHECK_EXECUTABLE)
 endif()
 
 if(CPPCHECK_EXECUTABLE)
+  # The reason for checking of the printed out file is empty is because
```
Comment (nilsdeppe) [inline cmake/SetupCppCheck.cmake]: of -> if

**Review by kidder — DISMISSED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 14]
```diff
@@ -18,24 +18,27 @@ At a high level, describe what this PR does.
...
+- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
+- Mark objects `const` whenever possible
+- Almost always `auto`, except with expression templates, i.e. `DataVector`
+- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
+- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
+- Prefix commits addressing PR requests with `fixup`. These commits are flagged
+  by Travis so we do not accidentally merge them.
+- Include what you use, prefer forward declarations in hpp files.
+- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
+- Use `Tensor`'s non-member get if the indices are constant expressions
```
Comment (fmahebert) [inline .github/PULL_REQUEST_TEMPLATE.md]: I think this could benefit from an example, perhaps "e.g. use `get<1>(tensor)`"

File: cmake/SetupCppCheck.cmake [hunk 15]
```diff
@@ -6,19 +6,41 @@ if(NOT CPPCHECK_FOUND AND NOT CPPCHECK_EXECUTABLE)
 endif()
 
 if(CPPCHECK_EXECUTABLE)
+  # The reason for checking if the printed out file is empty is because
```
Comment (fmahebert) [inline cmake/SetupCppCheck.cmake]: I prefer something along the lines of:
The reason for redirecting the output of cppcheck into the file `.cppcheck_output.txt` is
because cppcheck returns 0 even if some checks fail. Testing for a non-empty output file
is a more robust way of determining whether cppcheck found errors.

File: cmake/SetupCppCheck.cmake [hunk 16]
```diff
@@ -6,19 +6,41 @@ if(NOT CPPCHECK_FOUND AND NOT CPPCHECK_EXECUTABLE)
...
+  # cppcheck check returns 0 even if some checks fail. If the
+  # --error-exitcode flag is set, then it always returns the error
+  # code.
```
Comment (fmahebert) [inline cmake/SetupCppCheck.cmake]: It doesn't look like the `--error-exitcode` flag is set, so why is it discussed? If it's as "extra information", it should be made clear this sentence is not describing the actual code being run (so maybe mark as a note, or put in parens, or similar...).

File: docs/MainSite/Installation.md [hunk 17]
```diff
@@ -46,7 +46,11 @@ See LICENSE.txt for details.
 
 A [Docker](https://www.docker.com/) image is available from
 [DockerHub](https://hub.docker.com/r/sxscollaboration/spectrebuildenv/) and can
-be used to build SpECTRE on a personal machine.
+be used to build SpECTRE on a personal machine. If you have SELinux active
+on your system you must figure out how to enable sharing files with the host
+OS. If you receive errors that you do not have permission to access a shared
+directory it is likely that your system has SELinux enabled. One option is to
+disable SELinux at the expense of reducing the security of your system.
```
Comment (fmahebert) [inline docs/MainSite/Installation.md]: I like this clarification, but I feel that it could be moved to a note (like the note about not supporting all machines immediately below), or somehow be made less prominent? The clarification has become the majority of the paragraph that opens the Docker section.

I'm happy to be over-ruled on this.

File: src/Parallel/Algorithm.hpp [hunk 18]
```diff
@@ -169,6 +169,14 @@ void apply_visitor_helper(boost::variant<Variants...>& box,
           [&box](std::false_type /*returns_void*/, auto&&... my_args) {
             box = std::get<0>(Invokable::apply(boost::get<ThisVariant>(box),
                                                std::forward<Args>(my_args)...));
+            using return_box_type = decltype(std::get<0>(Invokable::apply(
+                boost::get<ThisVariant>(box), std::forward<Args>(my_args)...)));
+            static_assert(
+                cpp17::is_same_v<std::decay_t<return_box_type>, ThisVariant> or
+                    cpp17::is_same_v<db::DataBox<tmpl::list<>>, ThisVariant>,
+                "An explicit single Action must return either void, the same "
+                "DataBox that was passed into it, or only be invokable with an "
```
Comment (fmahebert) [inline src/Parallel/Algorithm.hpp]: I find it clearer to phrase as:
"... must return either void or the same DataBox that was pass into it, or must only be..."

English needs better scoping rules.

File: cmake/SetupCppCheck.cmake (hunk 15 above)
Comment (nilsdeppe) [reply cmake/SetupCppCheck.cmake]: But that's not why we redirect the output. We redirect the output because this way it's easier to spot the errors since they can be printed afterwards.

File: cmake/SetupCppCheck.cmake (hunk 16 above)
Comment (nilsdeppe) [reply cmake/SetupCppCheck.cmake]: Because cppcheck has that option and so if someone looks at what cppcheck offers they'd be confused why we aren't using that flag. This comment is designed to explain that

File: docs/MainSite/Installation.md (hunk 17 above)
Comment (nilsdeppe) [reply docs/MainSite/Installation.md]: Sounds good to me, I will make it a note

File: src/Parallel/Algorithm.hpp (hunk 18 above)
Comment (nilsdeppe) [reply src/Parallel/Algorithm.hpp]: Sure, I can reword this

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 14 above)
Comment (nilsdeppe) [reply .github/PULL_REQUEST_TEMPLATE.md]: Sure

File: cmake/SetupCppCheck.cmake (hunk 15 above)
Comment (fmahebert) [reply cmake/SetupCppCheck.cmake]: Fair enough. Then I would just suggest rewording "the printed out file" -> "the output file `.cppcheck_output.txt`" to be explicit.

File: cmake/SetupCppCheck.cmake (hunk 16 above)
Comment (fmahebert) [reply cmake/SetupCppCheck.cmake]: Then can I suggest "We do not use cppcheck's `--error-exitcode` flag, because this makes cppcheck always return the error code"

Comment (nilsdeppe) [reply cmake/SetupCppCheck.cmake]: sounds good to me

File: cmake/SetupCppCheck.cmake (hunk 15 above)
Comment (nilsdeppe) [reply cmake/SetupCppCheck.cmake]: Sure

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
